### PR TITLE
Rewrite the ISO8601 parser

### PR DESCRIPTION
**Pull request by gergocs**
The previous parser implementation didn't parse valid ISO8601 strings.

Signed-off-by: Gergo Csizi gergocs@inf.u-szeged.hu

### CODE REVIEW

File: src/builtins/BuiltinTemporal.cpp [hunk 1]
```diff
@@ -324,8 +324,11 @@ static Value builtinTemporalCalendarFrom(ExecutionState& state, Value thisValue,
 
 static Value builtinTemporalCalendarPrototypeId(ExecutionState& state, Value thisValue, size_t argc, Value* argv, Optional<Object*> newTarget)
 {
-    Value calendar = thisValue;
-    return calendar.asObject()->asTemporalCalendarObject()->getIdentifier();
+    if (!(thisValue.isObject() && thisValue.asObject()->isTemporalCalendarObject())) {
```
Comment (zherczeg) [inline src/builtins/BuiltinTemporal.cpp:327]: I think there is a helper for this check.

File: src/runtime/TemporalObject.cpp [hunk 2]
```diff
@@ -68,228 +68,284 @@ Value TemporalObject::toISOTime(ExecutionState& state, DateObject& d)
...
-                    year = "";
+    return { { "milliSecond", tmp.substr(0, 3) },
+             { "microSecond", tmp.substr(3, 3) },
+             { "nanoSecond", tmp.substr(6, 3) } };
+}
+
+void TemporalObject::offset(std::string& isoString)
+{
+    isoString = isoString.substr(1);
+    if (!(((isoString[0] == '0' || isoString[0] == '1') && isoString[1] >= '0' && isoString[1] <= '9') || (isoString[0] == '2' && isoString[1] >= 0 && isoString[1] <= '3'))) {
```
Comment (zherczeg) [inline src/runtime/TemporalObject.cpp:109]: The string is zero terminated, isn't it? So no need any length checks here.

File: src/runtime/TemporalObject.cpp [hunk 3]
```diff
@@ -68,228 +68,284 @@ Value TemporalObject::toISOTime(ExecutionState& state, DateObject& d)
...
-    ASSERT(str.isString());
-    size_t pos = str.asString()->toNonGCUTF8StringData().find('[');
-    std::string isoDate = str.asString()->toNonGCUTF8StringData().substr(0, pos);
-    std::string extra;
+    std::string retVal;
+    if (!TemporalObject::is_number(isoString.substr(0, n))) {
+        throw std::exception();
+    }
+    retVal = isoString.substr(0, n);
+    isoString = isoString.substr(n);
```
Comment (zherczeg) [inline src/runtime/TemporalObject.cpp:79]: A generic question: this method requires a lot of string  reallocation, which is bad for performance. Would it be possible to use an offset or char pointer?

File: src/runtime/TemporalObject.cpp [hunk 4]
```diff
@@ -68,228 +68,284 @@ Value TemporalObject::toISOTime(ExecutionState& state, DateObject& d)
...
-        } else if (date.length() > 4) {
-            if (date.find('W') != std::string::npos && date.length() == 5) {
-                goto good;
-            } else if (date.length() > 2) {
-                if (date[2] == '-' && date.length() == 5) {
-                    goto good;
+        }
+
+        // Calendar
+        if (isoString.length() != 0 && isoString.rfind("[u-ca=") == 0) {
```
Comment (zherczeg) [inline src/runtime/TemporalObject.cpp:283]: `isoString.length() > 6` isn't it?

**Review by zherczeg — CHANGES_REQUESTED** [review]

File: src/runtime/TemporalObject.cpp (hunk 2 above)
Comment (gergocs) [reply src/runtime/TemporalObject.cpp:109]: Yes it is.

File: src/runtime/TemporalObject.cpp (hunk 3 above)
Comment (gergocs) [reply src/runtime/TemporalObject.cpp:79]: Yes i will use offsets.

File: src/runtime/TemporalObject.cpp (hunk 4 above)
Comment (gergocs) [reply src/runtime/TemporalObject.cpp:283]: Yes

File: src/builtins/BuiltinTemporal.cpp (hunk 1 above)
Comment (gergocs) [reply src/builtins/BuiltinTemporal.cpp:327]: Yes but that function also checks if `argc` is not 0 and in this case `argc` will be 0.

**Comment by gergocs** [comment]
Patch is ready for review

File: src/runtime/TemporalObject.cpp [hunk 5]
```diff
@@ -68,228 +68,270 @@ Value TemporalObject::toISOTime(ExecutionState& state, DateObject& d)
...
+}
 
-    pos = isoDate.find('T');
-
-    if (pos != std::string::npos) {
-        std::replace(isoDate.begin(), isoDate.end(), 'T', ' ');
+std::map<std::string, int> TemporalObject::getSeconds(std::string& isoString, unsigned int& index)
+{
+    int counter = 1;
+    while (TemporalObject::is_number(isoString.substr(counter, 1))) {
```
Comment (zherczeg) [inline src/runtime/TemporalObject.cpp:86]: Is there a single character chekc for digits?

File: src/runtime/TemporalObject.cpp [hunk 6]
```diff
@@ -68,228 +68,270 @@ Value TemporalObject::toISOTime(ExecutionState& state, DateObject& d)
...
-    } else if (date.length() > 3) {
-        if (date.find('W') != std::string::npos && date.length() > 6) {
-            if (date.length() > 7) {
-                goto good;
+
+    TIME:
+        // Time
+        if (isoString[index] == 'T' || isoString.rfind("\\s", index) == 0) {
+            ++index;
+            for (; isoString[index] == 's'; ++index)
```
Comment (zherczeg) [inline src/runtime/TemporalObject.cpp:223]: `{}` missing

File: src/runtime/TemporalObject.cpp [hunk 7]
```diff
@@ -68,228 +68,270 @@ Value TemporalObject::toISOTime(ExecutionState& state, DateObject& d)
...
+            ++index;
+        } else if (isoString[index] == '+' || isoString[index] == '-' || isoString.rfind("−", index) == 0) {
+            TemporalObject::offset(isoString, index);
+        }
+
+
+        // TimeZone
+        if (isoString[index] != 'E') {
+            while (TemporalObject::tzComponent(isoString, index).length() != 0)
+                ;
```
Comment (zherczeg) [inline src/runtime/TemporalObject.cpp:260]: Is htis the preferred way? I thought `{}` must be used

File: src/runtime/TemporalObject.cpp [hunk 8]
```diff
@@ -68,228 +68,270 @@ Value TemporalObject::toISOTime(ExecutionState& state, DateObject& d)
...
+                goto TIME;
             }
+            throw std::exception();
         }
-    } else if (date.length() > 3) {
-        if (date.find('W') != std::string::npos && date.length() > 6) {
-            if (date.length() > 7) {
-                goto good;
+
+    TIME:
```
Comment (zherczeg) [inline src/runtime/TemporalObject.cpp:219]: Why capital letters?

**Review by zherczeg — CHANGES_REQUESTED** [review]

File: src/runtime/TemporalObject.cpp (hunk 5 above)
Comment (gergocs) [reply src/runtime/TemporalObject.cpp:86]: Yes `std::isdigit()`

**Comment by gergocs** [comment]
Patch is ready for review

File: src/builtins/BuiltinTemporal.cpp (hunk 1 above)
Comment (clover2123) [reply src/builtins/BuiltinTemporal.cpp:327]: (IMHO)
Wrapping the result of  AND(`&&`) operation with a NOT(`!`) operation seems not quite intuitive.
What about changing it to OR(`||`)?

File: src/runtime/TemporalObject.h [hunk 9]
```diff
@@ -35,14 +35,35 @@ static int monthDayLookUpTable[12] = { 31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 3
...
-    static std::map<std::string, std::string> parseTemporalDateString(ExecutionState& state, std::string isoString);
-    static std::map<std::string, std::string> parseTemporalDateTimeString(ExecutionState& state, String* isoString);
+    static TemporalObject::DateTime parseValidIso8601String(ExecutionState& state, std::string isoString);
+    static TemporalObject::DateTime parseTemporalDateString(ExecutionState& state, const std::string& isoString);
+    static TemporalObject::DateTime parseTemporalDateTimeString(ExecutionState& state, const std::string& isoString);
+    static std::string getNNumberFromString(std::string& isoString, int n, unsigned int& index);
+    static std::map<std::string, int> getSeconds(std::string& isoString, unsigned int& index);
+    static void offset(std::string& isoString, unsigned int& index);
+    static std::string tzComponent(std::string& isoString, unsigned int& index);
+    static bool is_number(const std::string& s)
```
Comment (clover2123) [inline src/runtime/TemporalObject.h:63]: Please change this method name to `isNumber` if possible
(We avoid using under bar in function naming)

File: src/runtime/TemporalObject.cpp [hunk 10]
```diff
@@ -68,228 +68,268 @@ Value TemporalObject::toISOTime(ExecutionState& state, DateObject& d)
...
 
-    if (!extra.empty()) {
-        size_t n = std::count(extra.begin(), extra.end(), '[');
-        if (n == 1 || n == 2) {
-            goto good;
+        if (isoString.length() != index) {
+            throw std::exception();
         }
+    } catch (std::exception& e) {
+        ErrorObject::throwBuiltinError(state, ErrorObject::RangeError, "Invalid ISO string");
```
Comment (clover2123) [inline src/runtime/TemporalObject.cpp:307]: It seems that the exception value (e) is not mandatory and nowhere used, so please replace this try-catch statement with simple if-else statement

File: src/runtime/TemporalObject.cpp [hunk 11]
```diff
@@ -68,228 +68,268 @@ Value TemporalObject::toISOTime(ExecutionState& state, DateObject& d)
...
-        if (pos == 0 && date.length() > 4) {
-            if (date.substr(pos + 1).find('-') == 0) {
-                date.erase(std::remove(date.begin(), date.end(), '-'), date.end());
-                if (date.length() == 4) {
-                    goto good;
+            if ((isoString[index] == '0' && std::isdigit(isoString[index + 1])) || (isoString[index] == '1' && isoString[index + 1] >= '0' && isoString[index + 1] <= '2')) {
+                dateTime.month = std::stoi(isoString.substr(index, 2));
+                index += 2;
+                if (isoString.empty()) {
+                    goto end;
```
Comment (clover2123) [inline src/runtime/TemporalObject.cpp:194]: I think that `goto` operation makes the code complex and difficult to understand, so please replace these goto operations as much as possible

**Comment by gergocs** [comment]
Patch is ready for review

**Review by clover2123 — APPROVED** [review]
LGTM

**Review by zherczeg — APPROVED** [review]
LGTM

**Pull request merged**